### PR TITLE
Add detailed client connection count metric.

### PR DESCRIPTION
**Pull request by OGKevin**
This metric adds the possibility to see how many front-end connections
are currently being used by user, db, client_host and hostgroup. This
makes it possible to add alerting for cases when a front-end client is
reaching its max connections.

### CODE REVIEW

**Comment by it-percona** [comment]
[CLA assistant check](https://cla-assistant.percona.com:443/percona/proxysql_exporter?pullRequest=25) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.percona.com:443/percona/proxysql_exporter?pullRequest=25) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.percona.com:443/check/percona/proxysql_exporter?pullRequest=25) it.</sub>

**Comment by OGKevin** [comment]
@AlekSi @rnovikovP may I get an ETA when someone is able to have a look at these PRs.

**Comment by OGKevin** [comment]
@AlekSi @rnovikovP ?

File: go.mod [hunk 1]
```diff
@@ -0,0 +1,31 @@
+module github.com/percona/proxysql_exporter
```
Comment (AlekSi) [inline go.mod]: Sorry, we can't migrate to Go modules at this point. Please use dep.

File: exporter.go [hunk 2]
```diff
@@ -29,25 +29,27 @@ const namespace = "proxysql"
...
+	scrapeDetailedMySQLProcessList bool
+	scrapesTotal                   prometheus.Counter
+	scrapeErrorsTotal              *prometheus.CounterVec
+	lastScrapeError                prometheus.Gauge
+	lastScrapeDurationSeconds      prometheus.Gauge
+	proxysqlUp                     prometheus.Gauge
 }
 
 // NewExporter returns a new ProxySQL exporter for the provided DSN.
 // It scrapes stats_mysql_global and stats_mysql_connection_pool if corresponding parameters are true.
```
Comment (AlekSi) [inline exporter.go]: Please update that comment too.

File: exporter.go [hunk 3]
```diff
@@ -408,5 +416,49 @@ func scrapeMySQLConnectionList(db *sql.DB, ch chan<- prometheus.Metric) error {
...
+				prometheus.BuildFQName(namespace, "processlist", m.name),
+				m.help,
+				[]string{"user", "db", "client_host", "hostgroup"},
+				nil,
+			),
+			m.valueType,
+			res.count,
+			res.user, res.db, res.clientHost, res.hostGroup,
+		)
+	}
```
Comment (AlekSi) [inline exporter.go]: We should check `rows.Err()` and log an error.

File: proxysql_exporter.go [hunk 4]
```diff
@@ -40,6 +40,7 @@ var (
 	mysqlStatusF         = flag.Bool("collect.mysql_status", true, "Collect from stats_mysql_global (SHOW MYSQL STATUS).")
 	mysqlConnectionPoolF = flag.Bool("collect.mysql_connection_pool", true, "Collect from stats_mysql_connection_pool.")
 	mysqlConnectionListF = flag.Bool("collect.mysql_connection_list", true, "Collect connection list from stats_mysql_processlist.")
+	mysqlDetailedConnectionListF = flag.Bool("collect.detailed.stats_mysql_processlist", true, "Collect detailed connection list from stats_mysql_processlist.")
```
Comment (AlekSi) [inline proxysql_exporter.go]: Please make it `false` by default for compatibility.

**Comment by AlekSi** [comment]
Looks good overall, thank you. But we can't switch vendoring tool and add new metrics by default at this moment. That will change in the future, but not now.

File: go.mod (hunk 1 above)
Comment (OGKevin) [reply go.mod]: 😭

Comment (AlekSi) [reply go.mod]: I know, I know… 😄

**Comment by OGKevin** [comment]
@AlekSi i think i've address all your comments 👍

File: exporter.go (hunk 3 above)
Comment (AlekSi) [reply exporter.go]: @OGKevin ^

Comment (OGKevin) [reply exporter.go]: https://github.com/percona/proxysql_exporter/pull/25/files#diff-557eb2a359922e8de5f18397fed0cd99R461 👁 

it this not what you meant ?

Comment (AlekSi) [reply exporter.go]: Sorry, missed that :)

Comment (OGKevin) [reply exporter.go]: np

**Comment by AlekSi** [comment]
Merged, thanks!

**Pull request merged**